testutil/server.go

```diff
@@ -100,14 +101,57 @@ func defaultServerConfig() *TestServerConfig {
 	}
 }
 
+// Ports allocation state.
+//
+// We let the kernel choose random port number, so there's a good
+// chance the port is unused. But kernel could allocate the same
+// port twice, if agent did not have time to start listening on it.
+//
+// There is no obvious way to defend from the other processes in
+// the system, which will use OS-assigned ports, but at least it
+// will reduce duplicated ports shared by the agents.
+var nmap = make(map[int]struct{})
+var nmapmu sync.Mutex
+
 // randomPort asks the kernel for a random port to use.
 func randomPort() int {
-	l, err := net.Listen("tcp", "127.0.0.1:0")
-	if err != nil {
-		panic(err)
+	makeport := func() (int, bool) {
```

Comment:
Hi @ybubnov 

This is the testutil server and should never be used for production uses. Pushing the ports into a map does not solve the issue though (and I've talked with DC about this). Global varies are not shared across multiple packages: https://gist.github.com/sethvargo/ebc696c9ed8f541a9887307115918181.

If more than one package imports this package, they both start with empty, independent maps. It's really a terrible problem that does not have a good solution 😦